auth/client/session.go

```diff
@@ -0,0 +1,337 @@
+
+	// SetRefreshToken sets the refresh token if none
+	// is provided for the given url and service
+	SetRefreshToken(realm *url.URL, service, token string)
+}
+
+// This is the minimum duration a token can last (in seconds).
+// A token must not live less than 60 seconds because older versions
+// of the Docker client didn't read their expiration from the token
+// response and assumed 60 seconds.  So to remain compatible with
+// those implementations, a token must live at least this long.
+const minimumTokenLifetimeSeconds = 60
+
+type TokenHandler struct {
+	creds     CredentialStore
+	transport http.RoundTripper
+
+	offlineAccess bool
+	forceOAuth    bool
+	clientID      string
+	scopes        []Scope
+
+	tokenLock       sync.Mutex
+	tokenCache      string
+	tokenExpiration time.Time
+}
+
+// Scope is a type which is serializable to a string
+// using the allow scope grammar.
+type Scope interface {
+	String() string
+}
+
+// RepositoryScope represents a token scope for access
+// to a repository.
+type RepositoryScope struct {
+	Repository string
+	Class      string
+	Actions    []string
+}
+
+// String returns the string representation of the repository
+// using the scope grammar
+func (rs RepositoryScope) String() string {
+	repoType := "repository"
+	if rs.Class != "" {
+		repoType = fmt.Sprintf("%s(%s)", repoType, rs.Class)
+	}
+	return fmt.Sprintf("%s:%s:%s", repoType, rs.Repository, strings.Join(rs.Actions, ","))
+}
+
+//// RegistryScope represents a token scope for access
+//// to resources in the registry.
+//type RegistryScope struct {
+//	Name    string
+//	Actions []string
+//}
+//
+//// String returns the string representation of the user
+//// using the scope grammar
+//func (rs RegistryScope) String() string {
+//	return fmt.Sprintf("registry:%s:%s", rs.Name, strings.Join(rs.Actions, ","))
+//}
+
+// TokenHandlerOptions is used to configure a new token handler
+type TokenHandlerOptions struct {
+	Transport   http.RoundTripper
+	Credentials CredentialStore
+
+	OfflineAccess bool
+	ForceOAuth    bool
+	ClientID      string
+	Scopes        []Scope
+}
+
+// NewTokenHandler creates a new AuthenicationHandler which supports
+// fetching tokens from a remote token server.
+func NewTokenHandler(transport http.RoundTripper, creds CredentialStore, clientID, scope string, actions ...string) *TokenHandler {
+	return &TokenHandler{
+		transport:     transport,
+		creds:         creds,
+		offlineAccess: false,
+		forceOAuth:    false,
+		clientID:      clientID,
+		//scopes: []Scope{
+		//	RepositoryScope{
+		//		Repository: scope,
+		//		Actions:    actions,
+		//	},
+		//},
+	}
+}
+
+func (th *TokenHandler) client() *http.Client {
+	return &http.Client{
+		Transport: th.transport,
+		Timeout:   15 * time.Second,
+	}
+}
+
+func (th *TokenHandler) Scheme() string {
+	return "bearer"
+}
+
+func (th *TokenHandler) AuthorizeRequest(params map[string]string, scopes ...string) (string, error) {
+	//th.tokenLock.Lock()
```

Comment:
There was a token cache (`map[something]something`) I removed that this was locking.